src/matching/Generic_vs_generic.ml

```diff
@@ -748,6 +753,13 @@ and m_expr ?(is_root = false) a b =
   | _, G.Alias (_alias, b1) -> m_expr a b1
   (* equivalence: user-defined equivalence! *)
   | G.DisjExpr (a1, a2), _b -> m_expr a1 b >||> m_expr a2 b
+  (* This case should only run exactly once.
+     This is so we do not endlessly loop trying to match to the same two things.
+     By setting `arguments_have_changed` to false, we ensure that we fall-through to
+     the cases that do decompose on `a` or `b`.
+  *)
+  | _, G.Cast (_, _, b1) when arguments_have_changed ->
+      m_expr a b1 >||> m_expr ~arguments_have_changed:false a b
```

Comment:
Wouldn't something like this work too?
```ocaml
  | a, B.Cast (b1, bt, b2) ->
      (match a with
      | G.Cast (a1, at, a2) ->
          m_type_ a1 b1 >>= fun () ->
          m_tok at bt >>= fun () -> m_expr a2 b2
      | a -> fail ())
      >||> m_expr a b1
```
It would be cleaner than this solution IMO.

Reply:
@brandonspark can you simpify the code like @IagoAbal asked? seems far simpler and more maintainable than having
yet another set of label arguments.

Reply:
It has been 5 days but I believe the idea was to replace the case
```ocaml
  | G.Cast (a1, at, a2), B.Cast (b1, bt, b2) -> ...
```
with something like what I proposed above. So you would try the `G.N (G.Id ((str, tok), _id_info)), _b when MV.is_metavar_name str` case before you tried the `a, B.Cast (b1, bt, b2)` case.

Reply:
im not necessarily sure i understand how you mean, i thought you were proposing to replace the new
```
| _, G.Cast (_, _, b1) when arguments_have_changed ->
      m_expr a b1 >||> m_expr ~arguments_have_changed:false a b
```
case (the third case) with 
```
  | a, B.Cast (b1, bt, b2) ->
      (match a with
      | G.Cast (a1, at, a2) ->
          m_type_ a1 b1 >>= fun () ->
          m_tok at bt >>= fun () -> m_expr a2 b2
      | a -> fail ())
      >||> m_expr a b1
```

if you meant to replace the `Cast _, Cast _` case with the thing above, then i think i misunderstood you. i dont see an issue with that i think, im fine with this.

Reply:
> if you meant to replace the Cast _, Cast _ case with the thing above, then i think i misunderstood you. i dont see an issue with that i think, im fine with this.

Yep I believe that's what I meant, I think it should work (but could be wrong) and if so it makes it easier to follow the logic.